### ISSUE TITLE
3.6.2 Detailed Design

### ISSUE THREAD

**Issue by kyusan93** [kyusan93/INF2001_P7-5#68]
Describe what are the possible design patterns (2 different patterns) you could incorporate in your design and provide a rationale for it. Show the template of your selected design patterns and then illustrates how you would implement the classes based on your class diagram. You do not need to include your whole class diagram, it can just be snippets of the relevant classes

**Comment by SelenaxPS** [comment]
The 2 design patterns that our team could incorporate into our design would be the Factory pattern and the Builder pattern.  

 

Factory Pattern 

Intent: The goal is to enhance adaptability in a system by defining an interface for creating objects in a superclass while allowing subclasses to modify the types of objects generated. 

Applicability: The Factory Pattern enhances the adaptability of our system by providing an interface for creating objects without specifying their exact types. In our workload management domain, we utilized the Factory Pattern to streamline the creation of users with different roles. By encapsulating the user creation logic within a factory, we can easily adjust and extend user roles, promoting flexibility to meet evolving system requirements. This approach shields the rest of the system from the intricacies of user creation and role assignment, fostering a modular and maintainable design. 
 

Builder Pattern 

Intent: The intent of the Builder pattern is to address issues associated with the construction of complex objects, allowing a more flexible and readable way to create instances with different configurations while encapsulating the construction logic.  

Applicability: The Builder pattern proves invaluable in constructing intricate objects, such as staff schedules and job assignments, within our workload management system. It significantly enhances code readability and simplifies the management of complex object construction. By employing the Builder pattern, we can systematically validate and enforce consistency at each stage of the construction process, mitigating the risk of creating objects in an incomplete or invalid state. 

For instance, we applied the Builder pattern to implement the creation of various planes and their corresponding flight details. This approach facilitated the separation of parameter setting and construction logic, leading to improved code readability. It also encapsulated the intricate details of plane and flight creation, contributing to a modular and maintainable system design.
